Integration tests flaking, timing out waiting for broker to come up

Opening a separate issue for this flake, as #643 started to become a general hub for all flaky behavior.

Integration tests are intermittently failing with error:

```
[Build & Unit/Integration Tests] --- FAIL: TestBasicFlows (40.09s)
[Build & Unit/Integration Tests]     controller_test.go:113: error waiting for broker to become ready: timed out waiting for the condition
```

```
[Build & Unit/Integration Tests] I0407 17:04:32.483384    3217 request.go:991] Response Body: {"kind":"Broker","apiVersion":"servicecatalog.k8s.io/v1alpha1","metadata":{"name":"test-broker","selfLink":"/apis/servicecatalog.k8s.io/v1alpha1/brokerstest-broker","uid":"33296edb-1bb4-11e7-8034-0242ac110002","resourceVersion":"18","creationTimestamp":"2017-04-07T17:04:02Z","finalizers":["kubernetes"]},"spec":{"url":"https://example.com"},"status":{"conditions":[{"type":"Ready","status":"False","reason":"ErrorSyncingCatalog","message":"Error syncing catalog from Broker"}]}}
[Build & Unit/Integration Tests] --- FAIL: TestBasicFlows (40.10s)
[Build & Unit/Integration Tests] 	controller_test.go:113: error waiting for broker to become ready: timed out waiting for the condition
[Build & Unit/Integration Tests] 	framework.go:58: Shutting down server on port: 37268
```
It'll be nice to have more errors in the conditions to have a bit more info.

Going by the `ErrorSyncingCatalog` we've got the catalog, but can't parse it, or can't reconcile it.

--- a/charts/catalog/README.md
+++ b/charts/catalog/README.md
@@ -56,7 +56,7 @@ chart and their default values.
 | `controllerManager.imagePullPolicy` | `imagePullPolicy` for the controller-manager | `Always` |
 | `controllerManager.verbosity` | Log level; valid values are in the range 0 - 10 | `10` |
 | `controllerManager.resyncInterval` | How often the controller should resync informers; duration format (`20m`, `1h`, etc) | `5m` |
-
+| `controllerManager.brokerRelistInterval` | How often the controller should relist the catalogs of ready brokers; duration format (`20m`, `1h`, etc) | `24h` |
 
 Specify each parameter using the `--set key=value[,key=value]` argument to
 `helm install`.

--- a/charts/catalog/templates/controller-manager-deployment.yaml
+++ b/charts/catalog/templates/controller-manager-deployment.yaml
@@ -44,6 +44,8 @@ spec:
         - "{{ .Values.controllerManager.verbosity }}"
         - --resync-interval
         - {{ .Values.controllerManager.resyncInterval }}
+        - --broker-relist-interval
+        - {{ .Values.controllerManager.brokerRelistInterval }}
         ports:
         - containerPort: 8080
         volumeMounts:

--- a/charts/catalog/values.yaml
+++ b/charts/catalog/values.yaml
@@ -54,3 +54,5 @@ controllerManager:
   verbosity: 10
   # Resync interval; format is a duration (`20m`, `1h`, etc)
   resyncInterval: 5m
+  # Broker relist interval; format is a duration (`20m`, `1h`, etc)
+  brokerRelistInterval: 24h

--- a/cmd/controller-manager/app/controller_manager.go
+++ b/cmd/controller-manager/app/controller_manager.go
@@ -274,7 +274,7 @@ func StartControllers(s *options.ControllerManagerServer,
 
 	// Launch service-catalog controller
 	if availableResources[schema.GroupVersionResource{Group: "servicecatalog.k8s.io", Version: "v1alpha1", Resource: "brokers"}] {
-		glog.V(5).Info("Creating shared informers")
+		glog.V(5).Info("Creating shared informers; resync interval: %v", s.ResyncInterval)
 		// Build the informer factory for service-catalog resources
 		informerFactory := servicecataloginformers.NewSharedInformerFactory(
 			serviceCatalogClientBuilder.ClientOrDie("shared-informers"),
@@ -283,7 +283,7 @@ func StartControllers(s *options.ControllerManagerServer,
 		// All shared informers are v1alpha1 API level
 		serviceCatalogSharedInformers := informerFactory.Servicecatalog().V1alpha1()
 
-		glog.V(5).Info("Creating controller")
+		glog.V(5).Info("Creating controller; broker relist interval: %v", s.BrokerRelistInterval)
 		serviceCatalogController, err := controller.NewController(
 			coreClient,
 			serviceCatalogClientBuilder.ClientOrDie(controllerManagerAgentName).ServicecatalogV1alpha1(),
@@ -292,6 +292,7 @@ func StartControllers(s *options.ControllerManagerServer,
 			serviceCatalogSharedInformers.Instances(),
 			serviceCatalogSharedInformers.Bindings(),
 			openservicebroker.NewClient,
+			s.BrokerRelistInterval,
 			s.OSBAPIContextProfile,
 		)
 		if err != nil {

--- a/cmd/controller-manager/app/options/options.go
+++ b/cmd/controller-manager/app/options/options.go
@@ -36,6 +36,7 @@ type ControllerManagerServer struct {
 }
 
 const defaultResyncInterval = 5 * time.Minute
+const defaultBrokerRelistInterval = 24 * time.Hour
 const defaultContentType = "application/json"
 const defaultBindAddress = "0.0.0.0"
 const defaultPort = 10000
@@ -54,6 +55,7 @@ func NewControllerManagerServer() *ControllerManagerServer {
 			K8sKubeconfigPath:            defaultK8sKubeconfigPath,
 			ServiceCatalogKubeconfigPath: defaultServiceCatalogKubeconfigPath,
 			ResyncInterval:               defaultResyncInterval,
+			BrokerRelistInterval:         defaultBrokerRelistInterval,
 			OSBAPIContextProfile:         defaultOSBAPIContextProfile,
 		},
 	}
@@ -69,5 +71,6 @@ func (s *ControllerManagerServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.ServiceCatalogAPIServerURL, "service-catalog-api-server-url", "", "The URL for the service-catalog API server")
 	fs.StringVar(&s.ServiceCatalogKubeconfigPath, "service-catalog-kubeconfig", s.ServiceCatalogKubeconfigPath, "Path to service-catalog kubeconfig")
 	fs.DurationVar(&s.ResyncInterval, "resync-interval", s.ResyncInterval, "The interval on which the controller will resync its informers")
+	fs.DurationVar(&s.BrokerRelistInterval, "broker-relist-interval", s.BrokerRelistInterval, "The interval on which a broker's catalog is relisted after the broker becomes ready")
 	fs.BoolVar(&s.OSBAPIContextProfile, "enable-osb-api-context-profile", s.OSBAPIContextProfile, "Whether or not to send the proposed optional OpenServiceBroker API Context Profile field")
 }

--- a/pkg/apis/componentconfig/types.go
+++ b/pkg/apis/componentconfig/types.go
@@ -57,6 +57,10 @@ type ControllerManagerConfiguration struct {
 	// all informers.
 	ResyncInterval time.Duration
 
+	// BrokerRelistInterval is the interval on which Broker's catalogs are re-
+	// listed.
+	BrokerRelistInterval time.Duration
+
 	// Whether or not to send the proposed optional
 	// OpenServiceBroker API Context Profile field
 	OSBAPIContextProfile bool

--- a/pkg/controller/controller.go
+++ b/pkg/controller/controller.go
@@ -52,6 +52,7 @@ func NewController(
 	instanceInformer informers.InstanceInformer,
 	bindingInformer informers.BindingInformer,
 	brokerClientCreateFunc brokerapi.CreateFunc,
+	brokerRelistInterval time.Duration,
 	osbAPIContextProfile bool,
 ) (Controller, error) {
 
@@ -67,6 +68,7 @@ func NewController(
 			brokerLister:              brokerLister,
 			serviceClassLister:        serviceClassLister,
 			instanceLister:            instanceLister,
+			brokerRelistInterval:      brokerRelistInterval,
 			enableOSBAPIContextProfle: osbAPIContextProfile,
 		}
 	)
@@ -113,6 +115,7 @@ type controller struct {
 	brokerLister              listers.BrokerLister
 	serviceClassLister        listers.ServiceClassLister
 	instanceLister            listers.InstanceLister
+	brokerRelistInterval      time.Duration
 	enableOSBAPIContextProfle bool
 }
 
@@ -159,9 +162,51 @@ const (
 	errorWithParameters         string = "ErrorWithParameters"
 )
 
+// shouldReconcileBroker determines whether a broker should be reconciled; it
+// returns true unless the broker has a ready condition with status true and
+// the controller's broker relist interval has not elapsed since the broker's
+// ready condition became true.
+func shouldReconcileBroker(broker *v1alpha1.Broker, now time.Time, relistInterval time.Duration) bool {
+	if broker.DeletionTimestamp != nil || len(broker.Status.Conditions) == 0 {
+		// If the deletion timestamp is set or the broker has no status
+		// conditions, we should reconcile it.
+		return true
+	}
+
+	// find the ready condition in the broker's status
+	for _, condition := range broker.Status.Conditions {
+		if condition.Type == v1alpha1.BrokerConditionReady {
+			// The broker has a ready condition
+
+			if condition.Status == v1alpha1.ConditionTrue {
+				// The broker's ready condition has status true, meaning that
+				// at some point, we successfully listed the broker's catalog.
+				// We should reconcile the broker (relist the broker's
+				// catalog) if it has been longer than the configured relist
+				// interval since the broker's ready condition became true.
+				return now.After(condition.LastTransitionTime.Add(relistInterval))
+			}
+
+			// The broker's ready condition wasn't true; we should try to re-
+			// list the broker.
+			return true
+		}
+	}
+
+	// The broker didn't have a ready condition; we should reconcile it.
+	return true
+}
+
 // reconcileBroker is the control-loop that reconciles a Broker.
 func (c *controller) reconcileBroker(broker *v1alpha1.Broker) {
 	glog.V(4).Infof("Processing Broker %v", broker.Name)
+
+	// If the broker's ready condition is true and the relist interval has not
+	// elapsed, do not reconcile it.
+	if !shouldReconcileBroker(broker, time.Now(), c.brokerRelistInterval) {
+		glog.V(10).Infof("Not processing Broker %v because relist interval has not elapsed since the broker became ready", broker.Name)
+		return
+	}
 
 	username, password, err := getAuthCredentialsFromBroker(c.kubeClient, broker)
 	if err != nil {

--- a/pkg/controller/controller_test.go
+++ b/pkg/controller/controller_test.go
@@ -186,6 +186,19 @@ func getTestBroker() *v1alpha1.Broker {
 	}
 }
 
+func getTestBrokerWithStatus(status v1alpha1.ConditionStatus) *v1alpha1.Broker {
+	broker := getTestBroker()
+	broker.Status = v1alpha1.BrokerStatus{
+		Conditions: []v1alpha1.BrokerCondition{{
+			Type:               v1alpha1.BrokerConditionReady,
+			Status:             status,
+			LastTransitionTime: metav1.NewTime(time.Now().Add(-5 * time.Minute)),
+		}},
+	}
+
+	return broker
+}
+
 // service class wired to the result of getTestBroker()
 func getTestServiceClass() *v1alpha1.ServiceClass {
 	return &v1alpha1.ServiceClass{
@@ -253,6 +266,83 @@ type instanceParameters struct {
 type bindingParameters struct {
 	Name string   `json:"name"`
 	Args []string `json:"args"`
+}
+
+func TestShouldReconcileBroker(t *testing.T) {
+	cases := []struct {
+		name      string
+		broker    *v1alpha1.Broker
+		now       time.Time
+		interval  time.Duration
+		reconcile bool
+	}{
+		{
+			name:      "no status",
+			broker:    getTestBroker(),
+			now:       time.Now(),
+			interval:  3 * time.Minute,
+			reconcile: true,
+		},
+		{
+			name: "no ready condition",
+			broker: func() *v1alpha1.Broker {
+				b := getTestBroker()
+				b.Status = v1alpha1.BrokerStatus{
+					Conditions: []v1alpha1.BrokerCondition{
+						{
+							Type:   v1alpha1.BrokerConditionType("NotARealCondition"),
+							Status: v1alpha1.ConditionTrue,
+						},
+					},
+				}
+				return b
+			}(),
+			now:       time.Now(),
+			interval:  3 * time.Minute,
+			reconcile: true,
+		},
+		{
+			name:      "not ready",
+			broker:    getTestBrokerWithStatus(v1alpha1.ConditionFalse),
+			now:       time.Now(),
+			interval:  3 * time.Minute,
+			reconcile: true,
+		},
+		{
+			name: "ready, interval elapsed",
+			broker: func() *v1alpha1.Broker {
+				broker := getTestBrokerWithStatus(v1alpha1.ConditionTrue)
+				return broker
+			}(),
+			now:       time.Now(),
+			interval:  3 * time.Minute,
+			reconcile: true,
+		},
+		{
+			name: "ready, interval not elapsed",
+			broker: func() *v1alpha1.Broker {
+				broker := getTestBrokerWithStatus(v1alpha1.ConditionTrue)
+				return broker
+			}(),
+			now:       time.Now(),
+			interval:  3 * time.Hour,
+			reconcile: false,
+		},
+	}
+
+	for _, tc := range cases {
+		var ltt *time.Time
+		if len(tc.broker.Status.Conditions) != 0 {
+			ltt = &tc.broker.Status.Conditions[0].LastTransitionTime.Time
+		}
+
+		t.Logf("%v: now: %v, interval: %v, last transition time: %v", tc.name, tc.now, tc.interval, ltt)
+		actual := shouldReconcileBroker(tc.broker, tc.now, tc.interval)
+
+		if e, a := tc.reconcile, actual; e != a {
+			t.Errorf("%v: unexpected result: expected %v, got %v", tc.name, e, a)
+		}
+	}
 }
 
 func TestReconcileBroker(t *testing.T) {
@@ -487,19 +577,6 @@ func TestReconcileBrokerWithReconcileError(t *testing.T) {
 }
 
 func TestUpdateBrokerCondition(t *testing.T) {
-	getTestBrokerWithStatus := func(status v1alpha1.ConditionStatus) *v1alpha1.Broker {
-		broker := getTestBroker()
-		broker.Status = v1alpha1.BrokerStatus{
-			Conditions: []v1alpha1.BrokerCondition{{
-				Type:               v1alpha1.BrokerConditionReady,
-				Status:             status,
-				LastTransitionTime: metav1.NewTime(time.Now().Add(-5 * time.Minute)),
-			}},
-		}
-
-		return broker
-	}
-
 	cases := []struct {
 		name                  string
 		input                 *v1alpha1.Broker
@@ -1866,6 +1943,7 @@ func newTestController(t *testing.T) (
 		serviceCatalogSharedInformers.Instances(),
 		serviceCatalogSharedInformers.Bindings(),
 		brokerClFunc,
+		24*time.Hour,
 		true,
 	)
 	if err != nil {

--- a/test/integration/controller_test.go
+++ b/test/integration/controller_test.go
@@ -275,6 +275,7 @@ func newTestController(t *testing.T) (
 		serviceCatalogSharedInformers.Instances(),
 		serviceCatalogSharedInformers.Bindings(),
 		brokerClFunc,
+		24*time.Hour,
 		true,
 	)
 	t.Log("controller start")

